File: 0115-distinct-subsequences/0115-distinct-subsequences.cpp
class Solution {
public:
    int Dis_sub(int i, int j, string &s1, string &s2, vector<vector<int>> &dp){
        if(j < 0) return 1;
        if(i < 0) return 0;

        if(dp[i][j] != -1) return dp[i][j];

        if(s1[i] == s2[j]){
            return dp[i][j] = Dis_sub(i-1, j-1, s1, s2, dp) + Dis_sub(i-1, j, s1, s2, dp);
        }

        return dp[i][j] = Dis_sub(i-1, j, s1, s2, dp);
    }

    int numDistinct(string s, string t) {
        int n = s.size();
        int m = t.size();
        vector<vector<int>> dp(n, vector<int>(m+1, -1));
        return Dis_sub(n-1, m-1, s, t, dp);
    }
};

File: 0115-distinct-subsequences/0115-distinct-subsequences.py
class Solution:
    def helper(self,i,j,str1, str2, dp):
        if j < 0:
            return 1
        if i < 0:
            return 0

        if dp[i][j] != -1:
            return dp[i][j]

        if str1[i] == str2[j]:
            dp[i][j] = self.helper(i-1, j-1, str1, str2, dp) + self.helper(i-1, j, str1, str2, dp)
            return dp[i][j]
        
        dp[i][j] = self.helper(i-1, j, str1, str2, dp)

        return dp[i][j]

    def numDistinct(self, s: str, t: str) -> int:
        n = len(s)
        m = len(t)
        dp = [[-1 for _ in range(m+1)] for _ in range(n+1)]
        return self.helper(n-1, m-1, s, t, dp)